ocaml/xapi/certificates.ml

```diff
@@ -282,12 +285,33 @@ let cert_perms kind =
   debug "%d %d" perm stat.Unix.st_perm ;
   perm
 
+let assert_valid kind ~name ~cert =
+  let cert =
+    match X509.Certificate.decode_pem (Cstruct.of_string cert) with
+    | Error (`Msg msg) ->
+        D.error "assert_valid: invalid cert. name='%s', msg='%s'" name msg ;
+        raise_corrupt kind name
+    | Ok cert ->
+        cert
+  in
+  let from', to' = X509.Certificate.validity cert in
+  let now = Ptime_clock.now () in
+  if Ptime.is_earlier ~than:from' now then (
+    D.error "assert_valid: cert becomes valid in the future... name='%s'" name ;
+    raise_corrupt kind name
+  ) else if Ptime.is_later ~than:to' now then (
+    D.error "assert_valid: cert has expired... name='%s'" name ;
+    raise_expired ~now:(Ptime.to_rfc3339 now) ~not_after:(Ptime.to_rfc3339 to')
+  ) else
+    ()
+
 let host_install kind ~name ~cert =
   if is_unsafe kind name then
     raise_name_invalid kind name ;
   let filename = library_filename kind name in
   if Sys.file_exists filename then
     raise_already_exists kind name ;
+  assert_valid kind ~name ~cert ;
```

Comment:
Can this check be done on on xapi_pool.ml in the certificate_install function? This forces the check to be done earlier, on the main host.

This also means that the check is _not_ performed for CRLs, which would fail because they cannot be parsed as normal certificates.